server-spi-private/src/main/java/org/keycloak/models/utils/DefaultKeyProviders.java

```diff
@@ -114,25 +114,53 @@ protected static boolean hasProvider(RealmModel realm, String providerId) {
 
     public static void createProviders(RealmModel realm, String privateKeyPem, String certificatePem) {
         if (!hasProvider(realm, "rsa")) {
-            ComponentModel rsa = new ComponentModel();
-            rsa.setName("rsa");
-            rsa.setParentId(realm.getId());
-            rsa.setProviderId("rsa");
-            rsa.setProviderType(KeyProvider.class.getName());
-
-            MultivaluedHashMap<String, String> config = new MultivaluedHashMap<>();
-            config.putSingle("priority", "100");
-            config.putSingle("privateKey", privateKeyPem);
-            if (certificatePem != null) {
-                config.putSingle("certificate", certificatePem);
-            }
-            rsa.setConfig(config);
-
-            realm.addComponentModel(rsa);
+            importRsaKEyProvider("rsa", realm, privateKeyPem, certificatePem);
+        }
+
+        if (!hasProvider(realm, "rsa-enc")) {
```

Comment:
I think that these codes are not needed because
- Both RSA key provider for signature and RSA key provider for encryption are created by using the same private key and certificate of a realm. Generally, it might not be appropriate to use the same key for several usage.
- This method is only called when importing a realm and the realm has its private key, public key and certificate. A realm's private key, public key and certificate are deprecated so that this method might be rarely called.

Reply:
Okay, I see where this method is calling, it's fine to me, it seems right.

But I have a question.

In the case of generated keys, it is creating for 2 rsa and encrypted rsa, see the link below. 

In this case it's right, because it generates another key and it's not the same key, right?

https://github.com/keycloak/keycloak/blob/7948d8d6e49a8588ff7389f5e52ee9e0bd4e827a/server-spi-private/src/main/java/org/keycloak/models/utils/DefaultKeyProviders.java#L40

Reply:
@fbrissi Yes. You are right.

In `AbstractGeneratedRsaKeyProviderFactory.validateConfiguration`
```
    @Override
    public void validateConfiguration(KeycloakSession session, RealmModel realm, ComponentModel model) throws ComponentValidationException {
        super.validateConfiguration(session, realm, model);

        ConfigurationValidationHelper.check(model).checkList(Attributes.KEY_SIZE_PROPERTY, false);

        int size = model.get(Attributes.KEY_SIZE_KEY, 2048);

        if (!(model.contains(Attributes.PRIVATE_KEY_KEY) && model.contains(Attributes.CERTIFICATE_KEY))) {
            generateKeys(realm, model, size);

            getLogger().debugv("Generated keys for {0}", realm.getName());
        } else {
            PrivateKey privateKey = PemUtils.decodePrivateKey(model.get(Attributes.PRIVATE_KEY_KEY));
            int currentSize = ((RSAPrivateKey) privateKey).getModulus().bitLength();
            if (currentSize != size) {
                generateKeys(realm, model, size);

                getLogger().debugv("Key size changed, generating new keys for {0}", realm.getName());
            }
        }
    }
```
In the case of generated RSA key, if clause is executed and a new rsa key is generated.
In the case of imported RSA key, else clause is executed and existing private key or certification in a composed model is used to get an rsa key.

Reply:
Reverted file

Reply:
Thanks for your review. 
His comments were very well explained.
I managed to do tests with imports and even see that in addition to being rare in the import I had done, the privateKey attribute is deprecated.